src/installer/managed/Microsoft.NET.HostModel/ResourceUpdater.cs

```diff
@@ -352,126 +182,219 @@ public ResourceUpdater AddResource(byte[] data, string lpType, IntPtr lpName)
+                        rsrcVirtualAddress);
+                    WriteI32(buffer, peSignatureOffset + Offsets.PEHeader.PE32DataDirectories + 2 * 8 + 4,
+                        rsrcSectionDataSize);
+                }
             }
+
+            buffer.WriteArray(rsrcPointerToRawData, rsrcSectionData, 0, rsrcSectionDataSize);
+
+            // clear rest
+            //Array.Fill is standard 2.1
+            for (int i = rsrcSectionDataSize; i < newSectionSize; i++)
+                buffer.Write(rsrcPointerToRawData + i, (byte)0);
+
+            _resourceData = null;
         }
 
-        private static void CaptureEnumResourcesErrorInfo(IntPtr errorInfoPtr)
+        private static int ReadI32(MemoryMappedViewAccessor buffer, int position)
         {
-            int hResult = Marshal.GetHRForLastWin32Error();
-            if (hResult != Kernel32.UserStoppedResourceEnumerationHRESULT)
-            {
-                GCHandle errorInfoHandle = GCHandle.FromIntPtr(errorInfoPtr);
-                var errorInfo = (EnumResourcesErrorInfo)errorInfoHandle.Target;
-                errorInfo.hResult = hResult;
-            }
+            return buffer.ReadByte(position + 0)
+                        | (buffer.ReadByte(position + 1) << 8)
+                        | (buffer.ReadByte(position + 2) << 16)
+                        | (buffer.ReadByte(position + 3) << 24);
         }
 
-        private sealed class ResourceNotAvailableException : Exception
+        private static void WriteI32(MemoryMappedViewAccessor buffer, int position, int data)
         {
-            public ResourceNotAvailableException(string message) : base(message)
-            {
-            }
+            buffer.Write(position + 0, (byte)(data & 0xFF));
+            buffer.Write(position + 1, (byte)(data >> 8 & 0xFF));
+            buffer.Write(position + 2, (byte)(data >> 16 & 0xFF));
+            buffer.Write(position + 3, (byte)(data >> 24 & 0xFF));
         }
-
-        private static void ThrowExceptionForLastWin32Error()
+        private static void WriteI16(MemoryMappedViewAccessor buffer, int position, short data)
         {
-            throw new HResultException(Marshal.GetHRForLastWin32Error());
+            buffer.Write(position + 0, (byte)(data & 0xFF));
+            buffer.Write(position + 1, (byte)(data >> 8 & 0xFF));
         }
 
-        private static void ThrowExceptionForInvalidUpdate()
+        private static void ModifyI32(MemoryMappedViewAccessor buffer, int position, Func<int, int> modifier) =>
+            WriteI32(buffer, position, modifier(ReadI32(buffer, position)));
+
+        public static int GetAligned(int integer, int alignWith) => (integer + alignWith - 1) & ~(alignWith - 1);
+
+        private static InvalidOperationException ThrowExceptionForInvalidUpdate()
         {
-            throw new InvalidOperationException("Update handle is invalid. This instance may not be used for further updates");
+            return new InvalidOperationException(
```

Comment:
Make this throw instead? It would would expect it to throw instead of return a function based the `Throw*` name.